README.md
Sycorax
=======

Sycorax is a Twitter client, written in Python, that choreographs the
online behavior of fictional characters. Other tweet schedulers make
your personal Twitter stream look like a clockwork robot is behind it,
posting tweets at the optimal time for penetration into your social
network. Syxorax lets fictional characters use Twitter the way real
people do. Your characters can post at odd hours and talk to each
other, taking their lines from a simple script you write, but without
any ongoing work from you.

For detailed instructions on scripting characters, see the project
home page at http://www.crummy.com/software/sycorax/


NEWS
= 20130627

* Put code on Github.
* Changed Sycorax to work with version 1.x of Python's 'twitter'
  module, which uses version 1.1 of the Twitter API.



tests.py
"""Unit tests."""

from datetime import datetime, timedelta
from unittest import main, TestCase
from timeline import TweetParser, Stream, Tweet, Day, Chapter
import pytz

# Begin mock objects.

class SycoraxTestCase(TestCase):

    AUTHORS = [
        dict(account="author1",
             code="",
             css_class="a1",
             color="red",
             ),

        dict(account="author2",
             code="+",
             css_class="a2",
             color="green",
             ),

        dict(account="author3",
             code="-",
             css_class="a3",
             color="blue",
             ),
        ]

    TIMEZONE = "US/Central"
    TIMEZONE_O = pytz.timezone(TIMEZONE)

    START_DATE = datetime(2000, 1, 1, 0, 0, 0, tzinfo=TIMEZONE_O)

    CONFIG = dict(authors=AUTHORS, timezone=TIMEZONE,
                  start_date=START_DATE, CHAPTER_DURATION=10)

    UID = 0

    @property
    def uid(self):
        v = str(self.UID)
        self.UID += 1
        return v

    def make_tweet(self, text=None, author=None, base_timecode=None,
                   timezone=None, delay=None, hour_of_day=None,
                   in_reply_to=None):
        text = text or self.uid
        author = author or self.AUTHORS[0]
        timezone = timezone or self.TIMEZONE
        base_timecode = base_timecode or self.START_DATE
        return Tweet(
            text, author, base_timecode, timezone, delay, hour_of_day,
            in_reply_to)

    def make_day(self, date=None, tweets=None):
        date = date or self.START_DATE
        return Day(date)

    def make_chapter(self, start_date=None):
        start_date = start_date or self.START_DATE
        return Chapter(start_date)

    def make_parser(self, config={}, fuzz_quotient=0, fuzz_minimum_seconds=0):
        base_config = dict(self.CONFIG)
        base_config.update(config)
        return TweetParser(base_config, fuzz_quotient, fuzz_minimum_seconds)

    def make_stream(self, tweet_parser=None, *lines):
        tweet_parser = tweet_parser or self.make_parser()
        return Stream(lines, tweet_parser)

    def assertDefaultAuthor(self, tweet):
        """Assert that the given tweet has the default author."""
        self.assertEquals(self.AUTHORS[0]['account'], tweet.author['account'])

    def assertDelayEquals(self, tweet, **kwargs):
        delay = timedelta(**kwargs)
        self.assertEquals(tweet.delay, delay)

class TestTweetParser(SycoraxTestCase):

    def tweet_for(self, line):
        stream = self.make_stream(None, line)
        return stream.latest_tweet

    def test_single_word_tweet(self):
        tweet = self.tweet_for("Foobar")
        self.assertEquals("Foobar", tweet.text)
        self.assertDefaultAuthor(tweet)

    def test_no_command(self):
        tweet = self.tweet_for("foo bar baz")
        self.assertEquals("foo bar baz", tweet.text)

    def test_tweet_that_looks_like_a_command_but_isnt(self):
        text = "Rh+ blood type"
        tweet = self.tweet_for(text)
        self.assertEquals(text, tweet.text)
        self.assertDefaultAuthor(tweet)

    def test_alternate_author(self):
        tweet = self.tweet_for("+ Foobar")
        self.assertEquals("Foobar", tweet.text)
        self.assertEquals(self.AUTHORS[1], tweet.author)

    def test_reply_to(self):
        stream = self.make_stream()
        t1 = stream.add_tweet("Original text")
        t2 = stream.add_tweet("R+ A reply")

        self.assertEquals(t2.in_reply_to, t1)

    def test_first_tweet_cannot_be_reply(self):
        self.assertRaises(ValueError, self.tweet_for, "R A reply")

    def test_delay_minutes(self):
        tweet = self.tweet_for("40M Foobar")
        self.assertEquals("Foobar", tweet.text)
        self.assertDelayEquals(tweet, minutes=40)

    def test_delay_days(self):
        tweet = self.tweet_for("2D Foobar")
        self.assertEquals("Foobar", tweet.text)
        self.assertDelayEquals(tweet, days=2)

    def test_hour_of_day_am(self):
        tweet = self.tweet_for("10A Foobar")
        self.assertEquals(10, tweet.hour_of_day)

    def test_hour_of_day_noon(self):
        tweet = self.tweet_for("12P Foobar")
        self.assertEquals(12, tweet.hour_of_day)

    def test_hour_of_day_pm(self):
        tweet = self.tweet_for("1P Foobar")
        self.assertEquals(13, tweet.hour_of_day)

    def test_hour_of_day_midnight(self):
        tweet = self.tweet_for("12A Foobar")
        self.assertEquals(0, tweet.hour_of_day)

    def test_multi_day_delay_plus_time_of_day(self):
        tweet = self.tweet_for("2D9A Foobar")
        self.assertDelayEquals(tweet, days=2)
        self.assertEquals(tweet.hour_of_day, 9)

    def test_less_than_day_delay_plus_time_of_day_fails(self):
        self.assertRaises(ValueError, self.tweet_for, "1H9A Foobar")

    def test_bad_time_of_day_fails(self):
        text = "13P Foobar"
        self.assertRaises(ValueError, self.tweet_for, text)

class TestTimecodeAssignment(SycoraxTestCase):

    def tweet_for(self, text=None, author=None, base_timecode=None,
              delay=None, hour_of_day=None, reply_to=None):

        tweet = self.make_tweet(
            text, author, base_timecode, self.TIMEZONE_O, delay, hour_of_day,
            reply_to)
        tweet.timestamp = tweet.calculate_timestamp(0, 0, None)
        return tweet

    def test_hour_of_day(self):
        tweet = self.tweet_for(hour_of_day=15)
        self.assertEquals(15, tweet.timestamp.hour)

    def test_long_delay_plus_hour_of_day(self):
        tweet = self.tweet_for(delay=timedelta(days=2), hour_of_day=5)
        self.assertEquals(tweet.timestamp.day, 3)
        self.assertEquals(tweet.timestamp.hour, 5)

    def test_hour_of_day_before_base_timecode_pushes_tweet_to_next_day(self):
        base = self.START_DATE.replace(hour=15)
        tweet = self.tweet_for(
            hour_of_day=10,
            base_timecode=base)
        self.assertEquals(tweet.timestamp.hour, 10)
        self.assertEquals(tweet.timestamp.day, base.day+1)

    def test_fuzz_on_relative_delay(self):
        parser = self.make_parser(fuzz_quotient=0.5)
        stream = self.make_stream(parser, "First tweet", "10M Second tweet")
        t1, t2 = stream.tweets
        self.assertEquals(self.START_DATE, t1.timestamp)

        # The second tweet's timestamp will not be precisely 10
        # minutes after the first, but it will be within 10 minutes,
        # plus or minus 50%.
        self.assertTrue(t2.timestamp-t1.timestamp < timedelta(minutes=15))

    def test_deterministic_tweets(self):
        parser = self.make_parser(fuzz_quotient=0, fuzz_minimum_seconds=0)
        stream = self.make_stream(parser, "First tweet", "1M Second tweet")
        t1, t2 = stream.tweets

        # The fuzz quotient and fuzz minimum seconds are zero, so the
        # timestamps are perfectly deterministic.
        self.assertEquals(t1.timestamp, self.START_DATE)
        self.assertEquals(t2.timestamp - t1.timestamp, timedelta(minutes=1))

    def test_minimum_fuzz_seconds(self):
        parser = self.make_parser(fuzz_quotient=0, fuzz_minimum_seconds=60)
        stream = self.make_stream(parser, "First tweet")
        [t1] = stream.tweets

        # The fuzz quotient is zero, so timestamps would be
        # deterministic, but there's a minimum of 60 seconds worth of
        # fuzz, so the first timestamp will be up to one minute off
        # from START_DATE.

        # This test will fail one time in 60, but we need to test that
        # *some* fuzz is being applied.
        self.assertTrue(t1.timestamp != self.START_DATE)
        difference = abs((t1.timestamp - self.START_DATE).seconds)
        self.assertTrue(difference < 60)

    def test_fuzz_on_hour_of_day(self):
        # A tweet that takes place in the ten o'clock hour will
        # take place sometime in the first 45 minutes of that hour.
        tweet = self.tweet_for(hour_of_day=10)

        # This will fail one time in 60, but it's important to check that
        # *some* fuzz is being applied.
        self.assertNotEquals(tweet.timestamp.minute, 0)
        self.assertTrue(tweet.timestamp.minute <= 45)

if __name__ == '__main__':
    main()


enact.py
"""Publish timelines to Twitter."""

from datetime import datetime, timedelta
from parsedatetime.parsedatetime import Calendar
import os
import json
import sys

import twitter

from timeline import load_config, JSON_TIME_FORMAT

from keys import TWITTER_CONSUMER_KEY, TWITTER_CONSUMER_SECRET

# This is a safety mechanism that prevents tweets that you modified by
# accident from being double-posted. If Sycorax stops running for some
# reason, this may also give you the opportunity to catch up on the
# tweets that were missed, rather than having an old tweet posted
# every time Sycorax runs.
DONT_POST_TWEETS_OLDER_THAN = timedelta(days=2)

TWITTER_TIME_FORMAT = "%a %b %d %H:%M:%S +0000 %Y"

class Story(object):

    def __init__(self, config, script_filehandle, progress_filename):
        self.progress_filename = progress_filename

        self.script = [json.loads(line.strip()) for line in script_filehandle]
        if os.path.exists(progress_filename):
            self.progress = [json.loads(line.strip())
                             for line in open(progress_filename)]
        else:
            # No progress yet
            self.progress = []
        self.credentials_by_account = {}
        for author in config['authors']:
            self.credentials_by_account[author['account']] = (
                author['twitter_token'], author['twitter_secret'])

        self.posted_tweets_by_internal_id = {}
        for posted_tweet in self.progress:
            self.posted_tweets_by_internal_id[posted_tweet['internal_id']] = (
                posted_tweet)

    def sync(self):
        """Synchronize by posting at most one tweet."""
        for tweet in self.script:
            if tweet['internal_id'] in self.posted_tweets_by_internal_id:
                # We already posted this tweet.
                continue
            else:
                # We have not yet posted this tweet.
                post_at = datetime.strptime(
                    tweet['timestamp'], JSON_TIME_FORMAT)
                now = datetime.utcnow()
                if post_at <= now:
                    if False and now - post_at > DONT_POST_TWEETS_OLDER_THAN:
                        print (
                            'Not posting "%s". It\'s so old (%s) that posting '
                            'it might screw up the timeline.' % (
                                tweet['text'], now-post_at))
                        continue
                    # It's time to post this sucker.
                    self.post(tweet)
                    continue
                else:
                    # This tweet's time has yet to come. Since the
                    # script is in chronological order, there's no
                    # point in looking further in the script.
                    print 'Coming up in %s: "%s"' % (post_at-now, tweet['text'])
                    pass
                break

    def post(self, tweet):
        text = tweet['text']
        print 'Posting "%s"' % text
        author_account = tweet['author']
        in_reply_to_id = tweet['in_reply_to']
        in_reply_to_twitter_id = None
        if in_reply_to_id is not None:
            # Find the actual Twitter ID of the tweet to which this is
            # a reply.
            in_reply_to = self.posted_tweets_by_internal_id.get(in_reply_to_id)
            if in_reply_to is None:
                print '"%s" is supposedly a response to nonexistent internal ID %s. Posting it as a standalone tweet instead.' % (text, in_reply_to_id)
            else:
                in_reply_to_twitter_id = in_reply_to['twitter_id']

        access_token_key, access_token_secret = self.credentials_by_account[
            author_account]
        oauth = twitter.OAuth(access_token_key, access_token_secret,
                              TWITTER_CONSUMER_KEY, TWITTER_CONSUMER_SECRET)
        api = twitter.Twitter(auth=oauth)

        # Post the tweet.
        try:
            data = api.statuses.update(status=text, in_reply_to_status_id=in_reply_to_twitter_id)
            actual_time = datetime.strptime(data['created_at'], TWITTER_TIME_FORMAT)
            twitter_id = data['id']
            pass
        except twitter.TwitterError, e:
            if e.message != "Status is a duplicate.":
                raise e
            actual_time = datetime.now()
            twitter_id = '[duplicate]'
        #actual_time = datetime.now()
        #twitter_id = 120943957396785

        # Append to the log of progress
        progress_entry = dict(
            text=text,
            planned_timestamp=tweet['timestamp'],
            actual_timestamp=actual_time.strftime(JSON_TIME_FORMAT),
            internal_id=tweet['internal_id'],
            twitter_id=twitter_id)
        self.progress.append(progress_entry)

        self.save_progress(progress_entry)

    def save_progress(self, entry):
        handle = open(self.progress_filename, "a")
        handle.write(json.dumps(entry))
        handle.write("\n")
        handle.close()

if len(sys.argv) != 2:
    print "Usage: %s [script directory]" % sys.argv[0]
    sys.exit()

script_directory = sys.argv[1]
config = load_config(script_directory)

script_filename = os.path.join(script_directory, "timeline.json")
if not os.path.exists(script_filename):
    raise Exception(
        "Could not find timeline.json file in directory %s. "
        "Did you run make_timeline.py?" % script_directory)

progress_filename = os.path.join(script_directory, "progress.json")

story = Story(config, open(script_filename), progress_filename)
story.sync()


timeline.py
"""Parse a Sycorax script into an annotated multi-author timeline."""

from datetime import datetime, timedelta
import json
import random
import re
import hashlib
import os
import pytz

# 10M: ~10 minutes later
# 4H: ~4 hours later
# 1D: the next day
# R: Reply to previous
# 10A: 10AM on the current day
# 9P: 9PM on the current day

DEFAULT_DELAY = timedelta(hours=4)
REPLY_TO_CODE = "R"
DELAY_CODE = re.compile("([0-9]+)([MHD])")
DELAY_UNITS = dict(M="minutes", H="hours", D="days")
TIME_OF_DAY_CODE = re.compile("([0-9]{1,2})([AP])")

JSON_TIME_FORMAT = "%d %b %Y %H:%M:%S %Z"

def load_config(directory):
    filename = os.path.join(directory, "config.json")
    if not os.path.exists(filename):
        raise Exception("Could not find config.json file in directory %s" % (
                directory
                ))

    data = json.loads(open(filename).read().strip())
    data['start_date'] = datetime.strptime(data['start_date'], "%Y/%m/%d")
    data['chapter_duration_days'] = timedelta(
        days=data['chapter_duration_days'])
    return data

def load_stream(directory):
    config = load_config(directory)
    filename = os.path.join(directory, "input.txt")
    if not os.path.exists(filename):
        raise Exception("Could not find input.txt file in directory %s" % (
                directory
                ))

    progress = load_progress(directory)
    try:
        progress = load_progress(directory)
    except Exception, e:
        # Nothing has been posted yet.
        progress = None

    return Stream(open(filename), config=config, progress=progress)


def load_progress(directory):
    config = load_config(directory)
    filename = os.path.join(directory, "progress.json")
    if not os.path.exists(filename):
        raise Exception("Could not find progress.json file in directory %s" % (
                directory
                ))
    return Progress(open(filename))


class TimezoneAware(object):

    def start_of_day(self, datetime):
        return datetime.replace(
            hour=0, minute=0, second=0, tzinfo=self.timezone)


class Progress(object):
    """The progress made in posting a stream."""

    def __init__(self, input_stream):
        self.timeline = [json.loads(line.strip()) for line in input_stream]
        self.posts = {}
        for post in self.timeline:
            self.posts[post['internal_id']] = post

class TweetParser(TimezoneAware):

    """Parses a line of script into a tweet."""

    def __init__(self, config, progress=None, fuzz_quotient=0.2,
                 fuzz_minimum_seconds=120):
        self.authors = config['authors']
        self.timezone = pytz.timezone(config['timezone'])
        for author in self.authors:
            author['account'] = author['account'].encode("utf8")
            author['css_class'] = author['account'].replace(
                '-', '').replace('_', '')
            author['color'] = author.get('color', 'white').encode("utf8")
            author['code'] = author.get('code', '')
        fuzz = float(config.get('fuzz', fuzz_quotient))
        self.fuzz_quotient = fuzz
        self.fuzz_minimum_seconds = int(config.get('fuzz_minimum_seconds', fuzz_minimum_seconds))
        self.start_date=config['start_date']
        self.config = config
        self.progress = progress

        self.default_author = None
        self.authors_by_code = {}
        for author in self.authors:
            code = author.get('code', '')
            if code == '':
                self.default_author = author
            self.authors_by_code[code] = author

    def parse(self, line, stream_so_far):
        is_command = False

        author = self.default_author
        reply_to = None
        delay = None
        hour_of_day = None

        if stream_so_far.latest_tweet is None:
            # This is the first tweet ever. The base timecode is the
            # start date.
            base_timecode = self.start_of_day(self.start_date)
        elif stream_so_far.current_chapter.total_tweets == 0:
            # This is the first tweet of the chapter. The base timecode
            # is the chapter start date.
            base_timecode = stream_so_far.current_chapter.start_date
        else:
            # There is no base timecode. The timestamp will be calculated
            # based on the previous tweet's timestamp.
            base_timecode = None

        line = line.strip()

        command_and_tweet = line.split(" ", 1)
        if len(command_and_tweet) > 1:
            command, tweet = command_and_tweet
        else:
            # Single-word tweet.
            return Tweet(line, author, base_timecode, self.timezone,
                         progress=self.progress)

        # The "command" may actually be the first word of the tweet.
        # Extract commands from it until there's nothing left.
        # If there is something left, it's not a command.
        for author_code, possible_author in self.authors_by_code.items():
            if author_code != "" and author_code in command:
                author = possible_author
                command = command.replace(author_code, "", 1)
                break

        is_reply = False
        if REPLY_TO_CODE in command:
            reply_to = stream_so_far.latest_tweet
            command = command.replace(REPLY_TO_CODE, "", 1)
            is_reply = True

        match = DELAY_CODE.match(command)
        if match is not None:
            number, unit = match.groups()
            subcommand = "".join(match.groups())
            command = command.replace(subcommand, "")
            kwargs = { DELAY_UNITS[unit]: int(number) }
            delay = timedelta(**kwargs)

        match = TIME_OF_DAY_CODE.match(command)
        if match is not None:
            hour, am = match.groups()
            subcommand = "".join(match.groups())
            command = command.replace(subcommand, "")
            hour = int(hour)
            if am == "A" and hour == 12:
                hour = 0
            if am == "P" and hour != 12:
                hour += 12
            if hour > 23:
                raise ValueError("Bad time of day %s in %s" % (
                        subcommand, line))
            hour_of_day = hour

        if command == "":
            # The first word has been entirely processed as
            # commands. The rest of the line is the actual content.
            line = tweet
        else:
            # The first word was not a command.
            author = self.default_author
            reply_to = None
            delay = None
            is_reply = False

        if is_reply and stream_so_far.latest_tweet is None:
            raise ValueError(
                "The first tweet in the script cannot be a reply.")

        if delay is None and hour_of_day is None:
            if len(stream_so_far.current_day.tweets) == 0 and stream_so_far.current_chapter.total_tweets > 0:
                # This is the first tweet of an in-story day, and no
                # special date instructions were given, so publish it at
                # the start of the next real-world day.
                base_timecode = self.start_of_day(base_timecode) + timedelta(
                    days=1)

            elif stream_so_far.current_chapter.total_tweets == 0:
                delay = timedelta(minutes=0)

        if len(line) > 140:
            print '[WARNING] %d characters in "%s"' % (len(line), line)
        return Tweet(line, author, base_timecode, self.timezone, delay,
                     hour_of_day, reply_to, self.progress)

class Chapter:

    def __init__(self, name, start_date):
        self.name = name
        self.days = []
        self.start_date = start_date

    @property
    def in_story_timeline_html(self):
        return "\n".join(
            ["<h2>%s</h2>\n" % self.name] +
            #["<p>%s days, %s tweets</p>\n" % (
            #        len(self.days), self.total_tweets)] +
            [day.in_story_timeline_html for day in self.days])

    @property
    def real_world_timeline_html(self):
        chapter_start_date = self.start_date.strftime(Tweet.REAL_WORLD_TIMELINE_DATE_FORMAT)
        if chapter_start_date != self.real_days[0].date:
            print '[WARNING] Chapter "%s" starts on %s, but its first tweet happens on %s' % (
                self.name, chapter_start_date, self.real_days[0].date)
        return "\n".join(
            ["<h2>%s</h2>\n" % self.name] +
            [day.real_world_timeline_html for day in self.real_days])


    @property
    def total_tweets(self):
        return sum(len(x.tweets) for x in self.days)

    @property
    def all_tweets(self):
        for d in self.days:
            for t in d.tweets:
                yield t

    @property
    def real_days(self):
        """A list of Day objects corresponding to real-world days for this chapter."""
        days = []
        current_date = None
        current_day = None
        for story_day in self.days:
            for tweet in story_day.tweets:
                if tweet.timestamp_date_str != current_date:
                    current_date = tweet.timestamp_date_str
                    current_day = Day(current_date)
                    days.append(current_day)
                current_day.tweets.append(tweet)
        return days


class Day:

    """A day's worth of tweets--either an in-story day or a real-world day."""

    def __init__(self, date):
        self.date = date
        self.tweets = []

    @property
    def in_story_timeline_html(self):
        if len(self.tweets) == 0:
            return ""
        return "\n".join(
            ["<h3>%s</h3>" % self.date, "<ul>"] +
            [tweet.in_story_timeline_html for tweet in self.tweets] + ["</ul>"])

    @property
    def real_world_timeline_html(self):
        if len(self.tweets) == 0:
            return ""
        return "\n".join(
            ["<h3>%s</h3>" % self.date, "<ul>"] +
            [tweet.real_world_timeline_html for tweet in self.tweets] + ["</ul>"])

class Tweet(TimezoneAware):

    REAL_WORLD_TIMELINE_TIME_FORMAT = "%H:%M"
    REAL_WORLD_TIMELINE_DATE_FORMAT = "%a %d %b"

    def __init__(self, text, author, base_timecode, timezone, delay=None,
                 hour_of_day=None, in_reply_to=None, progress=None):
        self.text = text
        self.author = author
        self.timezone = timezone
        self.in_reply_to = in_reply_to
        self.digest = hashlib.md5(self.text).hexdigest()
        self.delay = delay
        self.hour_of_day = hour_of_day
        if self.delay is None and self.hour_of_day is None:
            self.delay = DEFAULT_DELAY
        self.base_timecode = base_timecode

        # In general, timestamps are calculated in a second pass.
        self.timestamp = None

        # However, if this tweet has already been posted, we know its
        # timestamp already.
        if progress is not None:
            as_posted = progress.posts.get(self.digest)
            if as_posted is not None:
                self.timestamp = datetime.strptime(
                    as_posted['planned_timestamp'], JSON_TIME_FORMAT).replace(
                    tzinfo=pytz.timezone("UTC"))

        if (self.hour_of_day is not None and self.delay is not None
            and self.delay < timedelta(days=1)):
            raise ValueError(
                '"%s" defines both a delay and an hour of day, but the delay '
                'is less than one day.' % text)

    def calculate_timestamp(self, fuzz_quotient, fuzz_minimum_seconds,
                            previous_tweet):
        timestamp = self.base_timecode or previous_tweet.timestamp
        if self.timestamp is not None:
            # This tweet already has a timestamp, possibly because
            # it's already been posted. Leave it alone.
            return self.timestamp

        # If the delay after the last tweet is one day or more, apply
        # it before setting the time of day.
        one_day = timedelta(days=1)
        if self.delay is not None and self.delay >= one_day:
            timestamp += self.delay
            timestamp = self.start_of_day(timestamp)

        # If a time of day is given, set it now.
        if self.hour_of_day is not None:
            if timestamp.hour > self.hour_of_day:
                # Bump to the next real-world day.
                timestamp = timestamp + timedelta(days=1)
            timestamp = timestamp.replace(
                hour=self.hour_of_day, minute=0, second=0)

        # If the delay is less than one day, apply it now.
        if self.delay is not None and self.delay < one_day:
            timestamp += self.delay

        # Now we have a precise timestamp. But posting one tweet
        # exactly 30 minutes after another one will look fake. We need
        # to fudge the timestamp a little.

        if self.hour_of_day is not None:
            # We know which hour the tweet should go out. Pick
            # sometime in the first 45 minutes of that hour, to
            # minimize the chances of collisions with future tweets.
            actual_delta = timedelta(seconds=random.randint(0, 45*60))
            timestamp = timestamp + actual_delta
        elif self.delay is not None:
            # We know approximately how long after the previous tweet
            # this tweet should go out. Pick sometime
            delay_seconds = self.delay.seconds
            maximum_variation = max(
                delay_seconds * fuzz_quotient, fuzz_minimum_seconds)
            actual_variation = random.randint(-maximum_variation, maximum_variation)
            actual_delta = timedelta(seconds=actual_variation)
            if random.randint(0,1) == 1:
                timestamp = timestamp + actual_delta
            else:
                timestamp = timestamp - actual_delta
        else:
            raise ValueError(
                'Tweet "%s" has neither hour-of-day nor delay since previous '
                'tweet. Cannot calculate timestamp.' % self.text)
        return timestamp

    @property
    def json(self):
        if self.in_reply_to is None:
            in_reply_to = None
        else:
            in_reply_to = self.in_reply_to.digest
        d = dict(internal_id=self.digest, text=self.text,
                 author=self.author['account'],
                 in_reply_to=in_reply_to, timestamp=self.timestamp_for_json)
        return json.dumps(d)

    def li(self, text):
        a = []
        if self.in_reply_to is not None:
            a.append("<ul>")
        a.append('<li class="%s">%s</li>' % (self.author['css_class'], text))
        if self.in_reply_to is not None:
            a.append("</ul>")
        return "\n".join(a)

    @property
    def in_story_timeline_html(self):
        return self.li(self.text)

    @property
    def timestamp_str(self):
        return self.timestamp.strftime(self.REAL_WORLD_TIMELINE_TIME_FORMAT)

    @property
    def timestamp_for_json(self):
        return self.timestamp.astimezone(pytz.timezone("UTC")).strftime(
            JSON_TIME_FORMAT)

    @property
    def timestamp_date_str(self):
        return self.timestamp.strftime(self.REAL_WORLD_TIMELINE_DATE_FORMAT)

    @property
    def real_world_timeline_html(self):
        text = self.timestamp_str + " " + self.text
        return self.li(text)

class Stream:

    def __init__(self, lines, tweet_parser=None, config=None, progress=None):
        if tweet_parser is None:
            if config is None:
                raise ValueError(
                    "You tried to create a stream without providing a "
                    "tweet parser or a configuration for one.")
            tweet_parser = TweetParser(config=config, progress=progress)
        self.current_chapter = None
        self.current_day = None
        self.chapters = []
        self.tweet_parser = tweet_parser
        self.latest_tweet = None

        for line in lines:
            line = line.strip()
            if len(line) == 0:
                continue

            if line[:3] == "== ":
                self.end_chapter()
                self.begin_chapter(line[3:])
            elif line[:3] == "-- ":
                self.end_day()
                self.begin_day(line[3:])
            else:
                self.add_tweet(line)
        self.end_chapter()
        self.add_fuzz()
        self.chapter_start_sanity_check()

    def html_page(self, real_time=False):

        START = '''<html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8">
'''

        l = [START]
        l.append('<style type="text/css">')
        for author in self.tweet_parser.authors:
              l.append(".%s { background-color: %s }" % (
                      author['css_class'], author['color']))
        l.append('</style></head><body>')

        l.append("<p>Author guide:</p>")
        l.append("<ul>")
        for author in self.tweet_parser.authors:
            l.append('<li class="%s">%s</a>' % (author['css_class'], author['account']))
        l.append("</ul>")
        if real_time:
            l.append(self.real_world_timeline_html)
        else:
            l.append(self.in_story_timeline_html)
        l.append("</body></html")
        return "\n".join(l)

    @property
    def in_story_timeline_html(self):
        return "\n\n".join(chapter.in_story_timeline_html for chapter in self.chapters)

    @property
    def real_world_timeline_html(self):
        return "\n\n".join(chapter.real_world_timeline_html for chapter in self.chapters)

    @property
    def tweets(self):
        for chapter in self.chapters:
            for day in chapter.days:
                for tweet in day.tweets:
                    yield tweet

    def add_tweet(self, line):
        if self.current_chapter is None:
            self.begin_chapter("")
        if self.current_day is None:
            self.begin_day("")

        line = line.strip()
        tweet = self.tweet_parser.parse(line, self)
        self.current_day.tweets.append(tweet)
        self.latest_tweet = tweet
        return tweet

    def end_chapter(self):
        if self.current_chapter is None:
            # No current chapter.
            return

        self.end_day()
        self.current_chapter = None

    def begin_chapter(self, chapter_name):
        if len(self.chapters) == 0:
            start_date = self.tweet_parser.start_of_day(
                self.tweet_parser.config['start_date'])
        else:
            previous_chapter = self.chapters[-1]
            duration = self.tweet_parser.config['chapter_duration_days']
            start_date = previous_chapter.start_date + duration
        self.current_chapter = Chapter(chapter_name, start_date)
        self.chapters.append(self.current_chapter)

    def end_day(self):
        if self.current_day is None:
            return
        self.current_day = None

    def begin_day(self, date):
        self.current_day = Day(date)
        self.current_chapter.days.append(self.current_day)


    def add_fuzz(self):
        previous_tweet = None
        for tweet in self.tweets:
            progress = self.tweet_parser.progress
            if (progress is not None
                and progress.posts.get(tweet.digest) is not None):
                # This tweet has already been posted. Don't mess with it.
                previous_tweet = tweet
                continue
            success = False
            for i in range(0, 10):
                tweet.timestamp = tweet.calculate_timestamp(
                    self.tweet_parser.fuzz_quotient,
                    self.tweet_parser.fuzz_minimum_seconds,
                    previous_tweet)
                if (previous_tweet is None
                    or previous_tweet.timestamp < tweet.timestamp):
                    # This timestamp is fine. Stop trying to calculate it.
                    success = True
                    break
                # If we didn't break, the timestamp we calculated came
                # before previous tweet's timestamp, which is a
                # problem. Restart the loop and calculate a different
                # timestamp.
            if not success:
                # We tried to calculate the timestamp ten times with
                # no success. Raise an error.
                raise ValueError('Calculated timestamp for "%s" is %s, which comes before calculated timestamp for the previous tweet "%s" (%s). Trying again may help.' % (
                        tweet.text, tweet.timestamp_str, previous_tweet.text, previous_tweet.timestamp_str))
            previous_tweet = tweet

    def chapter_start_sanity_check(self):
        previous_chapter = self.chapters[0]
        for chapter in self.chapters[1:]:
            tweets = list(previous_chapter.all_tweets)
            if len(tweets) > 0:
                previous_chapter_last_tweet = tweets[-1]
                if previous_chapter_last_tweet.timestamp > chapter.start_date:
                    print '[WARNING] Last tweet in chapter "%s" overlaps the start of chapter "%s"' % (
                        previous_chapter.name, chapter.name)
            previous_chapter = chapter

    @property
    def json(self):
        return "\n".join(tweet.json for tweet in self.tweets)
